pkg/core/blockchain.go

```diff
@@ -613,18 +623,225 @@ func (bc *Blockchain) jumpToStateInternal(p uint32, stage stateJumpStage) error
+		return nil
+	}
+	if bc.config.KeepOnlyLatestState {
+		return fmt.Errorf("KeepOnlyLatestState is enabled, state for height %d is outdated and removed from the storage", height)
+	}
+	if bc.config.RemoveUntraceableBlocks && currHeight >= bc.config.MaxTraceableBlocks {
+		return fmt.Errorf("RemoveUntraceableBlocks is enabled, a necessary batch of traceable blocks has already been removed")
+	}
+
+	// Retrieve necessary state before the DB modification.
+	hHeight := bc.HeaderHeight()
+	b, err := bc.GetBlock(bc.headerHashes[height])
+	if err != nil {
+		return fmt.Errorf("failed to retrieve block %d: %w", height, err)
+	}
+	sr, err := bc.stateRoot.GetStateRoot(height)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve stateroot for height %d: %w", height, err)
+	}
+	v := bc.dao.Version
+	cache := bc.dao // dao is MemCachedStore over DB, so use dao directly to persist cached changes right to the underlying DB
+
+	bc.log.Info("initialize state reset", zap.Uint32("height", height))
+	start := time.Now()
+	p := start
+
+	resetStageKey := []byte{byte(storage.SYSStateChangeStage)}
+	switch stage {
+	case none:
+		cache.Store.Put(resetStageKey, []byte{stateResetBit | byte(stateJumpStarted)})
+		fallthrough
+	case stateJumpStarted:
+		// Remove headers/blocks/transactions/aers from currHeight down to height (not including height itself).
+		for i := height + 1; i <= hHeight; i++ {
+			err := cache.PurgeBlock(bc.headerHashes[i])
+			if err != nil {
+				return fmt.Errorf("error while removing block %d: %w", i, err)
+			}
+		}
+		cache.Store.Put(resetStageKey, []byte{stateResetBit | byte(staleBlocksRemoved)})
+
+		bc.log.Info("blocks, transactions ans AERs are reset", zap.Duration("duration", time.Since(p)))
+		p = time.Now()
+		fallthrough
+	case staleBlocksRemoved:
+		// Completely remove contract IDs to update them later.
+		cache.Store.Seek(storage.SeekRange{Prefix: []byte{byte(storage.STContractID)}}, func(k, _ []byte) bool {
+			cache.Store.Delete(k)
+			return true
+		})
+
+		// Reset contracts storage and store new contract IDs.
+		var mode = mpt.ModeAll
+		if bc.config.RemoveUntraceableBlocks {
+			mode |= mpt.ModeGCFlag
+		}
+		trieStore := mpt.NewTrieStore(sr.Root, mode, cache.Store)
+		oldStoragePrefix := v.StoragePrefix
+		newStoragePrefix := statesync.TemporaryPrefix(oldStoragePrefix)
+		mgmtCSPrefixLen := 1 + 4 + 1 // STStorage + Management ID + contract state prefix
+		mgmtContractPrefix := make([]byte, mgmtCSPrefixLen-1)
+		id := int32(native.ManagementContractID)
+		binary.BigEndian.PutUint32(mgmtContractPrefix, uint32(id))
+		mgmtContractPrefix[4] = native.PrefixContract
+		cs := new(state.Contract)
+
+		const persistBatchSize = 10000
+		var (
+			seekErr error
+			cnt     int
+		)
+		trieStore.Seek(storage.SeekRange{Prefix: []byte{byte(oldStoragePrefix)}}, func(k, v []byte) bool {
+			if cnt >= persistBatchSize {
+				_, seekErr = cache.Persist()
+				if seekErr != nil {
+					return false
+				}
+			}
+			cache.Store.Delete(k)
```

Comment:
I'm not sure why are you doing it here. We have some newer state stored at the `oldStoragePrefix`, it may have this `k`, it may not, it may have a lot of other keys, we need to seek through this prefix and drop everything at some later stage (`SeekGC` should work fine for this).

Reply:
It was implemented like this before the https://github.com/nspcc-dev/neo-go/pull/2773#discussion_r1018015248. But I've completely forgot that we use some previous storage state while seeking here. Of course, it's wrong and must not be done here.

Reply:
Fixed.